src/utils/order_state_utils.ts

```diff
@@ -0,0 +1,99 @@
+            const orderState: OrderStateValid = {
+                isValid: true,
+                orderHash,
+                orderRelevantState,
+            };
+            return orderState;
+        } catch (err) {
+            const orderState: OrderStateInvalid = {
+                isValid: false,
+                orderHash,
+                error: err.message,
+            };
+            return orderState;
+        }
+    }
+    public async getOrderRelevantStateAsync(
+        signedOrder: SignedOrder, methodOpts?: MethodOpts): Promise<OrderRelevantState> {
+        const zrxTokenAddress = await this.exchangeWrapper.getZRXTokenAddressAsync();
+        const orderHash = ZeroEx.getOrderHashHex(signedOrder);
+        const makerBalance = await this.tokenWrapper.getBalanceAsync(
+            signedOrder.makerTokenAddress, signedOrder.maker, methodOpts,
+        );
+        const makerProxyAllowance = await this.tokenWrapper.getProxyAllowanceAsync(
+            signedOrder.makerTokenAddress, signedOrder.maker, methodOpts,
+        );
+        const makerFeeBalance = await this.tokenWrapper.getBalanceAsync(
+            zrxTokenAddress, signedOrder.maker, methodOpts,
+        );
+        const makerFeeProxyAllowance = await this.tokenWrapper.getProxyAllowanceAsync(
+            zrxTokenAddress, signedOrder.maker, methodOpts,
+        );
+        const filledTakerTokenAmount = await this.exchangeWrapper.getFilledTakerAmountAsync(orderHash, methodOpts);
+        const canceledTakerTokenAmount = await this.exchangeWrapper.getCanceledTakerAmountAsync(orderHash, methodOpts);
+        const orderRelevantState = {
+            makerBalance,
+            makerProxyAllowance,
+            makerFeeBalance,
+            makerFeeProxyAllowance,
+            filledTakerTokenAmount,
+            canceledTakerTokenAmount,
+        };
+        return orderRelevantState;
+    }
+    private validateIfOrderIsValid(signedOrder: SignedOrder, orderRelevantState: OrderRelevantState): void {
+        const unavailableTakerTokenAmount = orderRelevantState.canceledTakerTokenAmount.add(
+            orderRelevantState.filledTakerTokenAmount,
+        );
+        const availableTakerTokenAmount = signedOrder.takerTokenAmount.minus(unavailableTakerTokenAmount);
+        if (availableTakerTokenAmount.eq(0)) {
+            throw new Error(ExchangeContractErrs.OrderRemainingFillAmountZero);
+        }
+
+        if (orderRelevantState.makerBalance.eq(0)) {
+            throw new Error(ExchangeContractErrs.InsufficientMakerBalance);
+        }
+        if (orderRelevantState.makerProxyAllowance.eq(0)) {
+            throw new Error(ExchangeContractErrs.InsufficientMakerAllowance);
+        }
+        if (!signedOrder.makerFee.eq(0)) {
+            if (orderRelevantState.makerFeeBalance.eq(0)) {
+                throw new Error(ExchangeContractErrs.InsufficientMakerFeeBalance);
+            }
+            if (orderRelevantState.makerFeeProxyAllowance.eq(0)) {
```

Comment:
It seems pretty gameable to only check for 0 balances/allowances, even though technically all orders are partially fillable with any balance/allowance.

Maybe we can let the user define some threshold (probably a percentage)?

Reply:
That is the idea! They will need to define a stricter definition within the callback. Our definition of validity in this context is the more relaxed definiton.